### PR TITLE
Nipype multiproc plugin modification to use GPU(s) as resources.

### PR DESCRIPTION
**Pull request by schahid**
Fixes # .

Changes proposed in this pull request
- Workflow nodes that run on GPUs can be scheduled similar to that of CPU resources. The plugin args will now require additionally "n_gpus" for total no. of GPUs to use and "ngpuproc" for no. of threads to start on a single GPU.

### CODE REVIEW

File: nipype/interfaces/base.py [hunk 1]
```diff
@@ -1570,10 +1570,12 @@ def __init__(self, command=None, terminal_output=None, **inputs):
         super(CommandLine, self).__init__(**inputs)
         self._environ = None
         # Set command. Input argument takes precedence
-        self._cmd = command or getattr(self, '_cmd', None)
-
-        if self._cmd is None:
+        if not hasattr(self, '_cmd'):
```
Comment (satra) [inline nipype/interfaces/base.py]: this bit should be reverted as the above command is equivalent.

File: nipype/pipeline/plugins/multiproc.py [hunk 2]
```diff
@@ -124,10 +135,60 @@ def __init__(self, plugin_args=None):
...
                                               get_system_total_memory_gb() * 0.9)
         self.raise_insufficient = self.plugin_args.get('raise_insufficient', True)
-
+        
+        self.n_gpus_visible = self.gpu_count()
+        self.n_gpus = self.plugin_args.get('n_gpus', self.n_gpus_visible)
+        self.n_gpu_proc = self.plugin_args.get('ngpuproc', 1)
+        
+        # Check plugin args
+        if self.plugin_args:
```
Comment (satra) [inline nipype/pipeline/plugins/multiproc.py]: i think this `if self.plugin_args` can be eliminated.

File: nipype/pipeline/plugins/multiproc.py [hunk 3]
```diff
@@ -124,10 +135,60 @@ def __init__(self, plugin_args=None):
...
+            self.n_gpus = self.n_gpus_visible
+            self.total_gpu_processors = self.n_gpus * self.n_gpu_proc
+        else:
+            #total gpu_processors = no.of GPUs * no.of threads per single GPU
+            self.total_gpu_processors = self.n_gpus * self.n_gpu_proc
+            
+
+        #form a GPU queue first
+        gpus=[]
+        try:
```
Comment (satra) [inline nipype/pipeline/plugins/multiproc.py]: this try catch is similar to gpu_count below. perhaps can be refactored to a common function.

**Comment by satra** [comment]
it would be good to merge this with master and resolve the conflicts.

**Comment by satra** [comment]
@schahid - is this something you can take care of? or should we take it over?

**Comment by schahid** [comment]
@satra - sorry for the delay in replying...as i am on holidays...yes sure you can take it over. thanks a lot.

**Comment by satra** [comment]
just putting a note here that with the reorganization of this module in release 1.0, this will require some significant merge effort.

**Comment by effigies** [comment]
Foolishly enough, I took a shot at that some time last week.

https://github.com/nipy/nipype/compare/master...effigies:enh/gpu_management

Since it's @schahid's `master` branch, I don't want to push to it directly. If anybody wants to take ownership of this PR, they can push that branch to their own repo, and open a new PR. (Or @schahid can `git merge` my branch into their `master` and resume work on it, .)

**Comment by satra** [comment]
@effigies  :) 

@schahid - would it be possible for you to take it on from @effigies changes?

**Comment by schahid** [comment]
Hi @satra and @effigies 
I have merged your branch into my master branch. Please let me know what else I should do?
Thanks.

**Pull request closed without merging**